pkg/storegateway/indexheader/thread.go

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/thread.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"runtime"
+)
+
+type execResult struct {
+	value interface{}
+	err   error
+	// populated is only set to `true` by the start method of OSThread. The field only
+	// exists so that we can tell when we've gotten a zero value of it. This is used by
+	// the Call method so that we can detect when the thread is shutting down.
+	populated bool
+}
+
+type OSThread struct {
+	// call is used to submit closures for this thread to execute.
+	call chan func() (interface{}, error)
+
+	// res is used to send the results of running closures back to callers.
+	res chan execResult
+
+	// stopping is a request for this thread to stop running.
+	stopping chan struct{}
+
+	// stopped is confirmation that the thread has stopped running.
+	stopped chan struct{}
+}
+
+func NewOSThread() *OSThread {
+	return &OSThread{
+		call:     make(chan func() (interface{}, error), 1),
+		res:      make(chan execResult, 1),
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+}
+
+func (o *OSThread) start() {
+	runtime.LockOSThread()
+
+	defer func() {
+		runtime.UnlockOSThread()
```

Comment:
I think it's safer to remove this.

Note that:
```go
// If the calling goroutine exits without unlocking the thread,
// the thread will be terminated.
```

And also:

```go
// Before calling UnlockOSThread, the caller must ensure that the OS
// thread is suitable for running other goroutines. If the caller made
// any permanent changes to the state of the thread that would affect
// other goroutines, it should not call this function and thus leave
// the goroutine locked to the OS thread until the goroutine (and
// hence the thread) exits.
```

Reply:
We don't do anything in the thread that would make it unsafe to be reused by other goroutines (the only code being run in them is code that was previously being run in plain goroutines) and so I'd rather not omit a cleanup method. However, I can do this just in case if people feel strongly.

Reply:
This isn't a strong opinion, I just think that it doesn't make sense to keep a code that does nothing (in this case).